woocommerce/woocommerce-gateway-stripe
# Add capture payment API endpoint

Fixes #2125

## Changes proposed in this Pull Request:

For Stripe Terminal support, the mobile apps need an API endpoint to trigger capturing of a [Stripe payment intent](https://stripe.com/docs/api/payment_intents). This PR adds a `wc_stripe/orders/<order_id>/capture_terminal_payment` endpoint to do so.

Additional Context: pdfdoF-er-p2

## Testing instructions

For making API requests, I found it helpful to install the [Basic Auth](https://github.com/WP-API/Basic-Auth) plugin to my test site so that I could make requests of the form `curl -X POST --user username:password 'http://YOUR_SITE/wp-json/wc/v3/wc_stripe/orders/<order_id>/capture_terminal_payment'` to test the endpoint.

### Test Case 1: Successful Payment Capture

1. Go to `/wp-admin/post-new.php?post_type=shop_order` to create a new order. Click **Add item(s)** followed by **Add product(s)**, and then add one or more products to the order. Click **Create** to create the order. Take note of both the Order ID and the Order Total.
2. Go to the Stripe Dashboard payments page (https://dashboard.stripe.com/test/payments) and click **+ Create payment**. Set the **Amount** to the order total from (1) and enter a Stripe test card number. Under **More options**, check the box that says **Capture funds later**. Click **Submit payment**. Once the payment is authorized, click **View details** to get to the payment page and take note of the Payment Intent ID (starts with `pi_`, located in the URL and in the upper right).
3. Run `curl -X POST --user USERNAME:PASSWORD 'http://YOUR_SITE/wp-json/wc/v3/wc_stripe/orders/ORDER_ID/capture_terminal_payment' --data "payment_intent_id=PAYMENT_INTENT_ID"`, using the Order ID from (1) and the Payment Intent ID from (2).
4. Verify that you get back a response with a `status` and `id`. The `status` should be `succeeded` and the `id` should be a Stripe Charge ID (starts with `ch_`). Take note of the Charge ID.
5. Go to the order page at `/wp-admin/post.php?post=ORDER_ID&action=edit`. Verify each of the following:
   - The order status should be "Completed".
   - The payment method title (after "Payment via") should be "WooCommerce Stripe In-Person Payments".
   - The Charge ID after the payment intent title should match the one you got in (3). Clicking should take you to the Stripe payment page where the payment should be labeled as "Succeeded".
6. Check the `wp_postmeta` table in the database for the order. Verify that the `_stripe_intent_id` value matches the Payment Intent ID from (2), the `payment_method` value is `stripe`, the `_stripe_charge_captured` value is `yes`, and the `_transaction_id` value is the Charge ID from (3).

### Test Case 2: Invalid Intent Status

1. Run `curl -X POST --user USERNAME:PASSWORD 'http://YOUR_SITE/wp-json/wc/v3/wc_stripe/orders/ORDER_ID/capture_terminal_payment' --data "payment_intent_id=PAYMENT_INTENT_ID"` again using the same Order ID and Payment Intent ID from Test Case 1.
2. This time, since the payment intent has already been captured, you should get a "The payment cannot be captured" error (status 409).

### Test Case 3: Refunded Orders

1. Go to `/wp-admin/post-new.php?post_type=shop_order` and create a new order with one or more products. Set the status to **Processing** and then **Create** the order. Then, click **Refund** and partially refund the order manually. Take note of the Order ID.
2. Go to the Stripe Dashboard payments page (https://dashboard.stripe.com/test/payments) and create a new payment for the net payment amount for the order. Be sure to check **Capture funds later** again. Take note of the Payment Intent ID.
3. Run `curl -X POST --user USERNAME:PASSWORD 'http://YOUR_SITE/wp-json/wc/v3/wc_stripe/orders/ORDER_ID/capture_terminal_payment' --data "payment_intent_id=PAYMENT_INTENT_ID"` using the Order ID from (1) and the Payment Intent ID from (2).
4. This should fail with a "Payment cannot be captured for partially or fully refunded orders." error (status 400).

### Test Case 4: Invalid Order ID

1. Run `curl -X POST --user USERNAME:PASSWORD 'http://YOUR_SITE/wp-json/wc/v3/wc_stripe/orders/ORDER_ID/capture_terminal_payment' --data "payment_intent_id=PAYMENT_INTENT_ID"` using an Order ID that doesn't exist in your store and the Payment Intent ID you used in Test Case 3 (or any other uncaptured Payment Intent ID).
3. Verify that you get an "Order not found" error (status 404).

### Test Case 5: Insufficient Authorized Amount

1. Create a new order with one or more products, take note of the Order Total and Order ID.
2. In your Stripe Dashboard, create a new payment (with **Capture funds later** checked) whose **Amount** is less than the Order Total from (1).
3. Run `curl -X POST --user USERNAME:PASSWORD 'http://YOUR_SITE/wp-json/wc/v3/wc_stripe/orders/ORDER_ID/capture_terminal_payment' --data "payment_intent_id=PAYMENT_INTENT_ID"` using the Order ID from (1) and the Payment Intent ID from (2).
4. Verify that you get a 502 error indicating that the requested capture amount is greater than the amount authorized by the payment intent.
---


**Post merge**
